pkg/entrypoint/entrypointer.go

```diff
@@ -467,3 +498,166 @@ func (e Entrypointer) outputRunResult(terminationReason string) result.RunResult
+	if template == "" {
+		return ArtifactTemplate{}, fmt.Errorf("template is empty")
+	}
+	if artifactref.StepArtifactOutputRegex.FindString(template) != template {
+		return ArtifactTemplate{}, fmt.Errorf("invalid artifact template %s", template)
+	}
+	template = strings.TrimSuffix(strings.TrimPrefix(template, "$("), ")")
+	split := strings.Split(template, ".")
+	at := ArtifactTemplate{
+
+		StepName: fmt.Sprintf("step-%s", split[1]),
+	}
+	if len(split) == 4 {
+		at.ArtifactName = split[3]
+	}
+	return at, nil
+}
+
+type ArtifactTemplate struct {
+	StepName     string
+	ArtifactName string
+}
+
+// applyStepArtifactSubstitutions replaces artifact references within a step's command and environment variables with their corresponding values.
+//
+// This function is designed to handle artifact substitutions in a script file, inline command, or environment variables.
+//
+// Args:
+//
+//	stepDir: The directory of the executing step.
+//	scriptDir: The directory where scripts are stored.
+//
+// Returns:
+//
+//	An error object if any issues occur during substitution.
+func (e *Entrypointer) applyStepArtifactSubstitutions(stepDir string, scriptDir string) error {
+
+	// Script was re-written into a file, we need to read the file to and substitute the content
+	// and re-write the command.
+	// While param substitution cannot be used in Script from StepAction, allowing artifact substitution doesn't seem bad as
+	// artifacts are unmarshalled, should be safe.
+	if len(e.Command) == 1 && filepath.Dir(e.Command[0]) == filepath.Clean(scriptDir) {
+		dataBytes, err := os.ReadFile(e.Command[0])
+		if err != nil {
+			return err
+		}
+		fileContent := string(dataBytes)
+		matches := artifactref.StepArtifactOutputRegex.FindAllStringSubmatch(fileContent, -1)
+		if len(matches) > 0 {
+			for _, m := range matches {
+				values, err := getArtifactValues(stepDir, m[0])
+				if err != nil {
+					return err
+				}
+				fileContent = strings.ReplaceAll(fileContent, m[0], values)
+			}
+			e.Command = []string{"sh", "-c", fileContent}
+		}
+	} else {
+		command := e.Command
+		newCmd := []string{}
+		for _, c := range command {
+			matches := artifactref.StepArtifactOutputRegex.FindAllStringSubmatch(c, -1)
+			if len(matches) > 0 {
+				for _, m := range matches {
+					values, err := getArtifactValues(stepDir, m[0])
+
+					if err != nil {
+						return err
+					}
+					c = strings.ReplaceAll(c, m[0], values)
+					newCmd = append(newCmd, c)
+				}
+			} else {
+				newCmd = append(newCmd, c)
+			}
+		}
+		e.Command = newCmd
+	}
+
+	// substitute env
+	for _, e := range os.Environ() {
+		pair := strings.SplitN(e, "=", 2)
+		matches := artifactref.StepArtifactOutputRegex.FindAllStringSubmatch(pair[1], -1)
+		for _, m := range matches {
+			values, err := getArtifactValues(stepDir, m[0])
+
+			if err != nil {
+				return err
+			}
+			err = os.Setenv(pair[0], strings.ReplaceAll(pair[1], m[0], values))
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
This block seems to be repeated twice. Can we wrap it in a function?

Reply:
I refactored this a little bit, cloud you take a look again when you have time, thanks.